### PR TITLE
T5Tokenizer adds EOS token if not already added

### LINKED ISSUE #5543: t5-base translation_en_to_de BLEU lower than the paper
I downloaded the "newstest2014.en" and "newstest2014.de" datasets. Then I used examples/translation/t5/evaluate_wmt.py to evaluate the BLEU value of en_to_de, and the BLEU finally obtained is equal to 22.15, which is much lower than the paper. I used the t5-base model and my transformers version is 2.11.0.
Is there something wrong with my operation? Is it necessary to fine-tune the t5 model to reproduce the results of the paper?

### LINKED ISSUE #5656: Truncated Outputs by t5 fine-tuned models
I fine-tuned t5-small over CNN/DM dataset using the finetune_t5.sh script. The outputs produced by the saved fine-tuned model is okayish but it's getting cut i.e., producing incomplete sentence at the end.

Example : 

Artcile:

(CNN)The only thing crazier than a guy in snowbound Massachusetts boxing up the powdery white stuff and offering it for sale online? People are actually buying it. For $89, self-styled entrepreneur Kyle Waring will ship you 6 pounds of Boston-area snow in an insulated Styrofoam box -- enough for 10 to 15 snowballs, he says.Kyle Waring died last week. But not if you live in New England or surrounding states. "We will not ship snow to any states in the northeast!" says Waring's website, ShipSnowYo.com. "We're in the business of expunging snow!" His website and social media accounts claim to have filled more than 133 orders for snow -- more than 30 on Tuesday alone, his busiest day yet. With more than 45 total inches, Boston has set a record this winter for the snowiest month in its history. Most residents see the huge piles of snow choking their yards and sidewalks as a nuisance, but Waring saw an opportunity. According to Boston.com, it all started a few weeks ago, when Waring and his wife were shoveling deep snow from their yard in Manchester-by-the-Sea, a coastal suburb north of Boston. He joked about shipping the stuff to friends and family in warmer states, and an idea was born. His business slogan: "Our nightmare is your dream!" At first, ShipSnowYo sold snow packed into empty 16.9-ounce water bottles for $19.99, but the snow usually melted before it reached its destination. So this week, Waring began shipping larger amounts in the Styrofoam cubes, which he promises will arrive anywhere in the U.S. in less than 20 hours. He also has begun selling a 10-pound box of snow for $119. Many of his customers appear to be companies in warm-weather states who are buying the snow as a gag, he said. Whether Waring can sustain his gimmicky venture into the spring remains to be seen. But he has no shortage of product. "At this rate, it's going to be July until the snow melts," he told Boston.com. "But I've thought about taking this idea and running with it for other seasonal items. Maybe I'll ship some fall foliage."

Summary produced by t5-small fine-tuned over CNN/DM : 

Kyle Waring will ship you 6 pounds of snow in an insulated Styrofoam box for $89 . The self-styled entrepreneur says he will not ship snow to any states in the northeast . Waring's website and social media accounts claim to have filled more than 133 orders for snow . "We're in the business of expunging snow!" Waring says . He has begun selling a 10-pound box of snow for $119 . His business slogan: "Our nightmare is your

At first I thought this might be because the model hasn't converged as I just ran for 1 epoch but it's producing similar truncated outputs even for t5-small fine-tuned over cnn/dm for 5 epochs.Also this problem is not related to min_length or max_length parameters I think, as it produced similar outputs for all combinations of those two parameters. Tried changing --max_source_length, --max_target_length, --val_max_target_length, --test_max_target_length(these 4 parameters are present in finetune.py) parameter's values too from their default values before fine-tuning but no use.

What might be the reason for this truncation? Is this a problem of the fine-tuning code used to fine-tune pretrained models as pre-trained models don't produce this kind of outputs.

### PR DESCRIPTION
**Pull request by sshleifer**
T5 Tokenizer should add `</s>` to the end of sequences. Since some users are doing this on their own, this PR only adds `</s>` if it has already been added.

On my machine, this makes zero shot validation BLEU go from 27.87 -> 27.65. Since this change is needed for finetuning, and the empirical difference is small and doesn't happen on Stas' machine, I would recommend merging this.

If others want to test, the command takes about 3 mins to run on brutasse.

### Zero Shot BLEU Scores
For english -> romanian
I grabbed the WMT english-romanian dataset:
```bash
wget https://s3.amazonaws.com/datasets.huggingface.co/translation/wmt_en_ro.tar.gz
tar -xzvf wmt_en_ro.tar.gz
```
Then ran evaluation (without finetuning) on the validation split:

```bash
export DATA_DIR=wmt_en_ro
python run_eval.py t5-base \
    $DATA_DIR/val.source t5_val_generations.txt \
    --reference_path $DATA_DIR/val.target \
    --score_path t5_enro_bleu_eos.json \
    --task translation_en_to_ro \
    --device cuda \
    --fp16 \
    --bs 32
```
(this branch) (with EOS):27.65
master (no EOS):  27.87
```
sacrebleu==1.4.3
torch==1.5.1
```
Will merge and fix tests if others have positive results.

### CODE REVIEW

**Comment by stas00** [comment]
FWIW, I get identical results of `27.84` with this branch and the master.

File: src/transformers/tokenization_t5.py [hunk 1]
```diff
@@ -145,6 +146,71 @@ def get_vocab(self):
...
+        self, token_ids_0: List[int], token_ids_1: Optional[List[int]] = None, already_has_special_tokens: bool = False
+    ) -> List[int]:
+        """
+        Retrieves sequence ids from a token list that has no special tokens added. This method is called when adding
+        special tokens using the tokenizer ``prepare_for_model`` method.
+
+        Args:
+            token_ids_0 (:obj:`List[int]`):
+                List of ids.
+            token_ids_1 (:obj:`List[int]`, `optional`, defaults to :obj:`None`):
```
Comment (sgugger) [inline src/transformers/tokenization_t5.py:159]: nit: no need to specify the default value if it's `None`.

File: src/transformers/tokenization_t5.py [hunk 2]
```diff
@@ -145,6 +146,71 @@ def get_vocab(self):
...
+                Optional second list of IDs for sequence pairs.
+            already_has_special_tokens (:obj:`bool`, `optional`, defaults to :obj:`False`):
+                Set to True if the token list is already formatted with special tokens for the model
+
+        Returns:
+            :obj:`List[int]`: A list of integers in the range [0, 1]: 1 for a special token, 0 for a sequence token.
```
Comment (sgugger) [inline src/transformers/tokenization_t5.py:165]: Not sure the second colon will please the mighty sphinx. Maybe a comma?
```suggestion
            :obj:`List[int]`: A list of integers in the range [0, 1], 1 for a special token, 0 for a sequence token.
```

**Review by sgugger — APPROVED** [review]
LGTM thanks! Just a few nits in the docs.

**Review by patrickvonplaten — APPROVED** [review]
Great! Very clean implementation. Not sure what to say about the evaluation. I think the functionality should be added anyways and the user can optionally set `add_special_tokens=False` if he/she wants

File: src/transformers/tokenization_t5.py [hunk 3]
```diff
@@ -148,6 +148,71 @@ def get_vocab(self):
...
+            return list(map(lambda x: 1 if x in [self.sep_token_id, self.cls_token_id] else 0, token_ids_0))
+        # normal case: some special tokens
+        if token_ids_1 is None:
+            return ([0] * len(token_ids_0)) + [1]
+        return ([0] * len(token_ids_0)) + [1] + ([0] * len(token_ids_1)) + [1]
+
+    def _add_eos_if_not_present(self, token_ids: List[int]) -> List[int]:
+        """Do not add eos again if user already added it."""
+        if len(token_ids) > 0 and token_ids[-1] == self.eos_token_id:
+            return token_ids
```
Comment (LysandreJik) [inline src/transformers/tokenization_t5.py:184]: Could we raise a warning/info telling the user that this is handled by the method, and that adding it manually + using the function would result in two </s> tokens being added in a future version?

**Review by LysandreJik — APPROVED** [review]
Great, nice addition. Maybe not the cleanest to check if there's already an existing </s> token as we don't do that in other implementations, but cool that this brings backwards-compatibility.

I believe we should align this with other tokenizers' prepare inputs methods in a future version, by always adding the EOS token even if there's already one in the input.

**Comment by sshleifer** [comment]
Why would it ever be good to add eos twice?

File: src/transformers/tokenization_t5.py (hunk 3 above)
Comment (sshleifer) [reply src/transformers/tokenization_t5.py:184]: sure.

**Comment by LysandreJik** [comment]
I think we can keep it like this right now with a warning for future versions. It would create a breaking change to users, and I feel it would be especially hard to debug an unknown drop in performance due to an additional token being added, right?

File: src/transformers/tokenization_t5.py (hunk 2 above)
Comment (sshleifer) [reply src/transformers/tokenization_t5.py]: done

**Comment by ahoho** [comment]
Will this behavior cause problems for the unsupervised setting? Per the [docs](https://huggingface.co/transformers/model_doc/t5.html#training), `</s>` is not added during denoising training:

```
input_ids = tokenizer.encode('The <extra_id_0> walks in <extra_id_1> park', return_tensors='pt')
labels = tokenizer.encode('<extra_id_0> cute dog <extra_id_1> the <extra_id_2> </s>', return_tensors='pt')
model(input_ids=input_ids, labels=labels)
```

Not sure if this will cause problems. (Also, Aa a somewhat related question, should the sentinel tokens in the `labels` be excluded from the loss in this setting, as I believe is the case with `[MASK]` in BERT?).

**Comment by sshleifer** [comment]
I'm not sure about either question:
made an issue verifying the docs: https://github.com/huggingface/transformers/issues/7904
Feel free to make an issue about the sentinel tokens question. I'd tag thomwolf/patrickvonplaten.
